ibis/spark/tests/test_udf.py

```diff
@@ -0,0 +1,373 @@
+    expr = fn(t.b, t.c)
+
+    assert isinstance(expr, ir.ColumnExpr)
+    assert isinstance(expr, ir.NumericColumn)
+    assert isinstance(expr, ir.FloatingColumn)
+
+    result = expr.execute()
+    expected = df.b + df.c
+    expected = Backend.default_series_rename(expected)
+    tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    'fn', my_add_fns
+)
+def test_multiple_argument_udf_group_by(con, t, df, fn):
+    expr = t.groupby(t.key).aggregate(
+        my_add=fn(t.b, t.c).sum()
+    ).sort_by('key')
+
+    assert isinstance(expr, ir.TableExpr)
+    assert isinstance(expr.my_add, ir.ColumnExpr)
+    assert isinstance(expr.my_add, ir.NumericColumn)
+    assert isinstance(expr.my_add, ir.FloatingColumn)
+
+    result = expr.execute()
+    expected = pd.DataFrame(
+        {'key': list('ab'), 'my_add': [sum([1.0 + 4.0, 2.0 + 5.0]), 3.0 + 6.0]}
+    )
+    tm.assert_frame_equal(result, expected)
+
+
+# Spark doesn't support pandas_udf GROUPED_AGG in spark.sql(). See SPARK-28422
+@pytest.mark.xfail
+def test_udaf(con, t, df):
+    expr = my_string_length_sum(t.a)
+
+    assert isinstance(expr, ir.ScalarExpr)
+
+    result = expr.execute()
+    expected = t.a.execute().str.len().mul(2).sum()
+    expected = Backend.default_series_rename(expected)
+    assert result == expected
+
+
+def test_udaf_groupby(con, t_random, df_random):
+    expr = t_random.groupby(t_random.key).aggregate(
+        my_corr=my_corr(t_random.a, t_random.b)
+    ).sort_by('key')
+
+    assert isinstance(expr, ir.TableExpr)
+
+    result = expr.execute()
+
+    dfi = df_random.set_index('key')
+    expected = pd.DataFrame(
+        {
+            'key': list('def'),
+            'my_corr': [
+                dfi.loc[value, 'a'].corr(dfi.loc[value, 'b'])
+                for value in 'def'
+            ],
+        }
+    )
+
+    tm.assert_frame_equal(result, expected)
+
+
+def test_nullable_output_not_allowed():
+    with pytest.raises(com.IbisTypeError):
+
+        d = dt.dtype('array<string>')
+        d.nullable = False
+
+        @udf.elementwise([dt.string], d)
+        def str_func(x):
+            pass
+
+
+def test_udaf_parameter_mismatch():
+    with pytest.raises(TypeError):
+
+        @udf.reduction(input_type=[dt.double], output_type=dt.double)
+        def my_corr(lhs, rhs, **kwargs):
+            pass
+
+
+def test_pandas_udf_zero_args_not_allowed():
+    with pytest.raises(com.UnsupportedArgumentError):
+
+        @udf.reduction(input_type=[], output_type=dt.double)
+        def my_corr2(**kwargs):
+            pass
+
+
+@pytest.mark.parametrize(
+    'times_two_fn', times_two_fns
+)
+@pytest.mark.parametrize(
+    'add_one_fn', add_one_fns
+)
+def test_compose_udfs(t_random, df_random, times_two_fn, add_one_fn):
+    expr = times_two_fn(add_one_fn(t_random.a))
+    result = expr.execute()
+    expected = df_random.a.add(1.0).mul(2.0)
+    expected = Backend.default_series_rename(expected)
```

Comment:
Should I use `ibis.tests.backends.Spark` or the `backend` fixture? Since `default_series_rename` is static, it might be cleaner to use the `Spark` class.

Reply:
Actually, yes, just use `Spark` instead of `Backend` (capital `B`).